keps/draft-20180418-security-profile.md

```diff
@@ -0,0 +1,1516 @@
+  templates:
+  - apiVersion: rbac.authorizatoin.k8s.io/v1
+    kind: Role
+    metadata:
+      name: use-psp
+    rules:
+    - apiGroups: ['policy']
+      resources: ['podsecuritypolicy']
+      verbs: ['use']
+      resourceNames: ['kickthetires-1.0-0']
+    - apiVersion: rbac.authorization.k8s.io/v1
+      kind: RoleBinding
+      metadata:
+        name: use-psp-binding
+      roleRef:
+        apiGroup: rbac.authorization.k8s.io/v1
+        kind: Role
+        name: use-psp
+      subjects:
+      - kind: Group
+        apiGroup: rbac.authorization.k8s.io/v1
+        name: system:serviceaccounts
+---
+apiVersion: profile.security.k8s.io/v1alpha1
+kind: SecurityProfile
+metadata:
+  name: kickthetires-1.0-0
+spec:
+  bootstrapping-rules:
+  - name: flags
+    spec:
+      apiserver:
+        anonymous-auth:
+          expect: unset
+        basic-auth-file:
+          expect: unset
+  - name: admission-controls
+    spec:
+      excludes:
+      - AlwaysAdmit
+  - name: authorization-mode
+    spec:
+      excludes:
+      - AlwaysAllow
+  runtime-rules:
+  - apiVersion: profile.security.k8s.io/v1alpha1
+    kind: NamespaceTemplate
+    name: kickthetires-1.0-0
+```
+
+A few curated security profiles can be defined to cover most common use cases and shared in the community, and helps use of Kubernetes:
+
+- Simplicity bringing up a secured cluster -- when a cluster admin specifies a security profile at cluster creation time,
+they can be assured that the cluster will provide the security and isolation properties they need,
+without them having to understand the low-level configuration used to achieve it.
+- The profile is sharable and reusable as it’s defined in a portable way
+- Enforced automatically and continuously
+
+Additionally, a Security Profile
+
+- Is versioned, to allow seamless upgrade of the cluster and Security Profile
+- Is highly customizable for particular needs
+- Is highly extendable with extra features
+
+An example if using security profile is
+
+```
+kubeadm init --security-profile=default-1.0-0
+```
+
+The command above will configure a secure cluster according to security profile `default-1.0-0`.
+
+As security profiles cover common/generic use cases,
+they are portable and reusable in most Kubernetes deployments,
+They are **NOT** defining policies specific to a particular Kubernetes deployment, like:
+
+- ResourceQuota
+- RBAC rules
+- Rules in NetworkPolicy for particular workloads
+
+The proposal also works with Config-as-Code systems,
+see the [section](#work-with-config-as-code-systems) below for details.
+
+### Proposed Security Profiles
+
+#### default
+
+Full example: [default-1.0-0](#the-default-profile)
+
+Minimum security practice to be compatible with most workloads.
+Best fit for sharing the cluster with trusted workloads.
+
+Enforcement:
+
+- etcd
+  - TLS required for communication
+  - Explicitly specify `--wal-dir`, and `--max-wals=0`
+- kube-apiserver
+  - TLS required for communication
```

Comment:
how is this enforced/ verified?

Reply:
They are defined in bootstrapping rules, the enforcement of bootstrapping rules are implementation specific, thus out of scope in this proposal.
Verification is out of scope.
I will call out these explicitly.